### PR TITLE
PUBDEV-5720 - xgboost: cross-validation doesn't work

### PR DESCRIPTION
**Pull request by Pscheidl**
https://0xdata.atlassian.net/browse/PUBDEV-5720

Prediction completely ignored weights.

### CODE REVIEW

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 1]
```diff
@@ -163,17 +164,11 @@ private static int sumChunksLength(int[] chunkIds, Vec vec, Vec weightsVector) {
...
-            int nzIndex = 0;
-            do {
-                nzIndex = weightVecChunk.nextNZ(nzIndex, true);
-                if (nzIndex < 0 || nzIndex >= weightVecChunk._len) break;
-                if (weightVecChunk.atd(nzIndex) == 0) totalChunkLength--;
-            } while (true);
+            if (weightsVector == null) {
+                totalChunkLength += vec.chunkLen(chunk);
+            } else {
+                totalChunkLength += countNonZerosInChunk(weightsVector.chunkForChunkIdx(chunk));
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: This part is not directly related to the issue. However, counting zeros in a chunk emerged on yet another place, so the code has been extracted and only called from this method.

If there are no weights, let's just add the size of the whole chunk to the overall length. If there are weights present, rows with weights of zero should not be counted.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 2]
```diff
@@ -323,7 +317,13 @@ private static DMatrix dense(Chunk[] chunks, int weight, DataInfo di, int respId
         float[][] data = allocateDenseMatrix(chunks[0].len(), di);
 
         long actualRows = denseChunk(data, chunks, weight, respIdx, di, cols, resp, weights);
-        assert actualRows == chunks[0].len();
+        if (weight == -1) {
+            assert actualRows == chunks[0].len();
+        } else {
+            // Rows with weights zero are skipped, thus the resulting number of rows must be equal
+            // to number of non-zero weights in weights chunk
+            assert actualRows == weights.length;
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: What about ? `assert actualRows == di.getWeightsVec().nzCnt()` here ? This `di.getWeightsVec().nzCnt()` method returns different number then `weights.length` in some cases ,which is really strange and the assert fails in such case. The `weights` array is allocated after manually counting number of non-zero elements in the weight vector - this way the assert always works. 

I wonder if I'm missing something.

**Comment by Pscheidl** [comment]
snimek z 2018-07-15 14-48-48

**Comment by Pscheidl** [comment]
Added JUnit tests.

**Comment by Pscheidl** [comment]
Fixed & enhanced test. I now test the scoring history for all cv frames for binomial, regression & multinomial.

File: h2o-py/tests/testdir_algos/xgboost/pyunit_H2OXGBoost_native_XGBoost_mixed_binomial_compare_large_dense_NOPASS_multinode.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def comparison_test_dense():
 
     # train the native XGBoost
     nativeTrain = pyunit_utils.convertH2OFrameToDMatrix(trainFile, y, enumCols=enumCols)
-    nativeModel = xgb.trainxgb.train(params=nativeParam,
```
[the remark this answers is not in the document]
Comment (Pscheidl) [reply h2o-py/tests/testdir_algos/xgboost/pyunit_H2OXGBoost_native_XGBoost_mixed_binomial_compare_large_dense_NOPASS_multinode.py]: NOPASS ?

[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/tests/testdir_algos/xgboost/pyunit_H2OXGBoost_native_XGBoost_mixed_binomial_compare_large_dense_NOPASS_multinode.py]: All tests are running in single node mode in order to pass.  Hence, they have been running all these time thanks to Michal R.  However, if you do not put NOPASS in the name, it will be called in general running mode and it will fail.

File: h2o-extensions/xgboost/src/main/java/ml/dmlc/xgboost4j/java/XGBoostScoreTask.java [hunk 4]
```diff
@@ -166,7 +180,7 @@ public void map(Chunk[] cs, NewChunk[] ncs) {
                 double[] dpreds = new double[preds.length];
                 for (int j = 0; j < dpreds.length; ++j)
                     dpreds[j] = preds[j][0];
-                for (int i = 0; i < cs[0]._len; ++i) {
+                for (int i = 0; i < (weightColumnIndex != -1 ? weights.length : cs[0].len()); ++i) {
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/main/java/ml/dmlc/xgboost4j/java/XGBoostScoreTask.java]: This looks suspicious, all the chunks should have the same length. Why do you need to take the `weights` as the reference?

File: h2o-extensions/xgboost/src/main/java/ml/dmlc/xgboost4j/java/XGBoostScoreTask.java [hunk 5]
```diff
@@ -137,11 +142,20 @@ public void map(Chunk[] cs, NewChunk[] ncs) {
...
+            } else {
+                weightColumnIndex = -1;
+            }
+
+            _sharedmodel._dataInfoKey.get().weightChunkId();
             data = XGBoostUtils.convertChunksToDMatrix(
                     _sharedmodel._dataInfoKey,
                     cs,
                     _fr.find(_parms._response_column),
-                    -1, // not used for preds
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/main/java/ml/dmlc/xgboost4j/java/XGBoostScoreTask.java]: Why is this wrong? You don't need weights to make predictions. It is not necessary to make predictions for rows with zero weight but it doesn't matter in terms of correctness. What matters is that the weights are accounted for in the MetricBuilder - which is true - MB is skipping predictions with zero weights.

**Comment by michalkurka** [comment]
I do not understand how this fixes cross validation issues. All other algos are making predictions even for observations with zero weights.

Users actually do expect to get predictions even for 0-weighted rows. I don't think this PR is correct.

**Comment by michalkurka** [comment]
@Pscheidl I think the actual issue is that Metrics are not calculated with weights in mind. This doesn't seem to be correct. Lets discuss tomorrow.

**Comment by Pscheidl** [comment]
Taking tests from this PR and closing. Will create a new one.
